README.md
Chef Wordpress Cookbook
=======================
The Chef Wordpress cookbook installs and configures Wordpress according to the instructions at http://codex.wordpress.org/Installing_WordPress.

This cookbook does not set up the WordPress blog. You will need to do this manually by going to http://hostname/wp-admin/install.php (this URL may be different if you change the attribute values).


Installation
------------
Install the cookbook using knife:

    $ knife cookbook site install wordpress

Or, if you are using Berkshelf, add the cookbook to your Berksfile:

```ruby
cookbook 'wordpress'
```


Usage
-----
Add the cookbook to your `run_list` in a node or role:

```json
{
  "run_list": [
    "recipe[wordpress::default]"
  ]
}
```

Or include it in a recipe:

```ruby
# other_cookbook/metadata.rb
# ...
depends 'wordpress'
```

```ruby
# other_cookbook/recipes/default.rb
# ...
include_recipe 'wordpress::default'
```

If a different version than the default is desired, download that version and get the SHA256 checksum (sha256sum on Linux systems), and set the version and checksum attributes.


Attributes
----------
`node['wordpress']` attributes:

<table>
  <thead>
    <tr>
      <th>Attribute</th>
      <th>Description</th>
      <th>Example</th>
      <th>Default</th>
    </tr>
  </thead>

  <tbody>
    <tr>
      <td>version</td>
      <td>version of the wordpress to install</td>
      <td><tt>1.2.3</tt></td>
      <td><tt>'latest'</tt></td>
    </tr>
    <tr>
      <td>checksum</td>
      <td>sha256sum of the tarball</td>
      <td><tt>abcd1234</tt></td>
      <td><tt>''</tt></td>
    </tr>
    <tr>
      <td>dir</td>
      <td>location for wordpress files</td>
      <td><tt>/nfs/wp</tt></td>
      <td><tt>/var/www</tt></td>
    </tr>
    <tr>
      <td>server_aliases</td>
      <td>server aliases for Apache</td>
      <td><tt>['foo.com']</tt></td>
      <td><tt>[(node's FQDN)]</tt></td>
    </tr>
  </tbody>
</table>

`node['wordpress']['db']` attributes:

<table>
  <thead>
    <tr>
      <th>Attribute</th>
      <th>Description</th>
      <th>Example</th>
      <th>Default</th>
    </tr>
  </thead>

  <tbody>
    <tr>
      <td>database</td>
      <td>name of the database to use</td>
      <td><tt>bob-wordpress</tt></td>
      <td><tt>wordpressdb</tt></td>
    </tr>
    <tr>
      <td>user</td>
      <td>the user to connect to MySQL</td>
      <td><tt>user</tt></td>
      <td><tt>wordpressuser</tt></td>
    </tr>
    <tr>
      <td>password</td>
      <td>the password to connect to MySQL</td>
      <td><tt>P@s$w0rD</tt></td>
      <td><tt>(randomly generated)</tt></td>
    </tr>
  </tbody>
</table>

`node['wordpress']['languages']` attributes:

<table>
  <thead>
    <tr>
      <th>Attribute</th>
      <th>Description</th>
      <th>Example</th>
      <th>Default</th>
    </tr>
  </thead>

  <tbody>
    <tr>
      <td>lang</td>
      <td>WPLANG value</td>
      <td><tt>ja</tt></td>
      <td><tt>''</tt></td>
    </tr>
    <tr>
      <td>version</td>
      <td>version of the <a href="http://translate.wordpress.org/projects/wp">WordPress translation projects</a></td>
      <td><tt>3.6.x</tt></td>
      <td><tt>''</tt></td>
    </tr>
    <tr>
      <td>projects</td>
      <td>translation project names to install</td>
      <td><tt>['main']</tt></td>
      <td><tt>['main', 'admin', 'admin_network', 'continents_cities']</tt></td>
    </tr>
    <tr>
      <td>themes</td>
      <td>theme translation project names to install</td>
      <td><tt>['twentytwelve']</tt></td>
      <td><tt>[]</tt></td>
    </tr>
  </tbody>
</table>


Attributes will probably never need to change (these all default to randomly generated strings):

* `node['wordpress']['keys']['auth']`
* `node['wordpress']['keys']['secure_auth']`
* `node['wordpress']['keys']['logged_in']`
* `node['wordpress']['keys']['nonce']`


Development
-----------
This cookbook uses Test Kitchen (1.0). To run the tests, clone the repository, install the gems, and run test kitchen:

    $ git clone git://github.com/opscode-cookbooks/wordpress.git
    $ cd wordpress
    $ bundle install
    $ bundle exec strainer test

1. Fork the cookbook on GitHub
2. Make changes
3. Write appropriate tests
4. Submit a Pull Request back to the project
5. Open a [JIRA ticket](https://tickets.opscode.com), linking back to the Pull Request


License
-------

Copyright:: 2015, Cloudenablers

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
CHANGELOG
=========

v1.2.0
------
Initial release


metadata.rb
name             "wordpress"
maintainer       "Cloudenablers"
maintainer_email "nagalakshmi.n@cloudenablers.com"
license          "Apache 2.0"
description      "Installs/Configures WordPress"
long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version          "1.2.0"

recipe "WordPress", "Installs and configures WordPress LAMP stack on a single system"
recipe "WordPress::languages", "Install WordPress translation files"

%w{ php openssl }.each do |cb|
  depends cb
end

depends "apache2", ">= 0.99.4"
depends "mysql", ">= 1.0.5"
depends "build-essential"

%w{ debian ubuntu }.each do |os|
  supports os
end

attribute "WordPress/version",
  :display_name => "WordPress download version",
  :description => "Version of WordPress to download from the WordPress site or 'latest' for the current release.",
  :default => "latest"

attribute "WordPress/checksum",
  :display_name => "WordPress tarball checksum",
  :description => "Checksum of the tarball for the version specified.",
  :default => ""

attribute "WordPress/dir",
  :display_name => "WordPress installation directory",
  :description => "Location to place WordPress files.",
  :default => "/var/www/wordpress"

attribute "WordPress/db/database",
  :display_name => "WordPress MySQL database",
  :description => "WordPress will use this MySQL database to store its data.",
  :default => "wordpressdb"

attribute "WordPress/db/user",
  :display_name => "WordPress MySQL user",
  :description => "WordPress will connect to MySQL using this user.",
  :default => "wordpressuser"

attribute "WordPress/db/password",
  :display_name => "WordPress MySQL password",
  :description => "Password for the WordPress MySQL user.",
  :default => "randomly generated"

attribute "WordPress/keys/auth",
  :display_name => "WordPress auth key",
  :description => "WordPress auth key.",
  :default => "randomly generated"

attribute "WordPress/keys/secure_auth",
  :display_name => "WordPress secure auth key",
  :description => "WordPress secure auth key.",
  :default => "randomly generated"

attribute "WordPress/keys/logged_in",
  :display_name => "WordPress logged-in key",
  :description => "WordPress logged-in key.",
  :default => "randomly generated"

attribute "WordPress/keys/nonce",
  :display_name => "WordPress nonce key",
  :description => "WordPress nonce key.",
  :default => "randomly generated"

attribute "WordPress/server_aliases",
  :display_name => "WordPress Server Aliases",
  :description => "WordPress Server Aliases",
  :default => "FQDN"

attribute "WordPress/languages/lang",
  :display_name => "WordPress WPLANG configulation value",
  :description => "WordPress WPLANG configulation value",
  :default => ""

attribute "WordPress/languages/version",
  :display_name => "Version of WordPress translation file",
  :description => "Version of WordPress translation file",
  :default => ""

attribute "WordPress/languages/projects",
  :display_name => "WordPress translation projects",
  :description => "WordPress translation projects",
  :type => "array",
  :default => ["main", "admin", "admin/network", "cc"]


recipes/database.rb
#
# Cookbook Name:: wordpress
# Recipe:: database
#
# Copyright (C) 2015, Cloudenablers
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe "mysql::client" unless platform_family?('windows') # No MySQL client on Windows

::Chef::Recipe.send(:include, Opscode::OpenSSL::Password)
::Chef::Recipe.send(:include, Wordpress::Helpers)

node.set_unless['wordpress']['db']['pass'] = secure_password
node.save unless Chef::Config[:solo]

db = node['wordpress']['db']

if is_local_host? db['host']
  include_recipe "mysql::server"

  mysql_bin = (platform_family? 'windows') ? 'mysql.exe' : 'mysql'
  user = "'#{db['user']}'@'#{db['host']}'"
  create_user = %<CREATE USER #{user} IDENTIFIED BY '#{db['pass']}';>
  user_exists = %<SELECT 1 FROM mysql.user WHERE user = '#{db['user']}';>
  create_db = %<CREATE DATABASE #{db['name']};>
  db_exists = %<SHOW DATABASES LIKE '#{db['name']}';>
  grant_privileges = %<GRANT ALL PRIVILEGES ON #{db['name']}.* TO #{user};>
  privileges_exist = %<SHOW GRANTS FOR for #{user}@'%';>
  flush_privileges = %<FLUSH PRIVILEGES;>

  execute "Create WordPress MySQL User" do
    action :run
    command "#{mysql_bin} #{::Wordpress::Helpers.make_db_query("root", node['mysql']['server_root_password'], create_user)}"
    only_if { `#{mysql_bin} #{::Wordpress::Helpers.make_db_query("root", node['mysql']['server_root_password'], user_exists)}`.empty? }
  end

  execute "Grant WordPress MySQL Privileges" do
    action :run
    command "#{mysql_bin} #{::Wordpress::Helpers.make_db_query("root", node['mysql']['server_root_password'], grant_privileges)}"
    only_if { `#{mysql_bin} #{::Wordpress::Helpers.make_db_query("root", node['mysql']['server_root_password'], privileges_exist)}`.empty? }
    notifies :run, "execute[Flush MySQL Privileges]"
  end

  execute "Flush MySQL Privileges" do
    action :nothing
    command "#{mysql_bin} #{::Wordpress::Helpers.make_db_query("root", node['mysql']['server_root_password'], flush_privileges)}"
  end

  execute "Create WordPress Database" do
    action :run
    command "#{mysql_bin} #{::Wordpress::Helpers.make_db_query("root", node['mysql']['server_root_password'], create_db)}"
    only_if { `#{mysql_bin} #{::Wordpress::Helpers.make_db_query("root", node['mysql']['server_root_password'], db_exists)}`.empty? }
  end
end


libraries/helpers.rb
#
# Cookbook Name:: wordpress
# Library:: helpers
#
# Copyright 2015, Cloudenablers
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

module Wordpress
  module Helpers
    def is_local_host?(host)
      if host == 'localhost' || host == '127.0.0.1' || host == '::1'
        true
      else
        require 'socket'
        require 'resolv'
        Socket.ip_address_list.map { |a| a.ip_address }.include? Resolv.getaddress host
      end
    end

    def self.make_db_query(user, pass, query)
      %< --user=#{user} --password="#{pass}" --execute="#{query}">
    end
  end
end


recipes/languages.rb
#
# Cookbook Name:: wordpress
# Recipe:: languages
#
# Copyright 2015, Cloudenablers
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe "wordpress"

directory "#{node['wordpress']['dir']}/wp-content/languages" do
  owner "root"
  group "root"
  mode "0755"
  action :create
  recursive true
end

unless node['wordpress']['languages']['lang'].to_s.empty? &&
       node['wordpress']['languages']['version'].to_s.empty?
  urls = node['wordpress']['languages']['urls']
  node['wordpress']['languages']['projects'].to_a.each do |project|
    next unless urls[project]

    file = "#{node['wordpress']['dir']}/wp-content/languages/"
    file += "#{project.tr('_', '-')}-" if project != 'main'
    file += "#{node['wordpress']['languages']['lang']}.mo"

    remote_file file do
      source urls[project]
      owner "root"
      group "root"
      mode "0644"
      action :create_if_missing
    end
  end

  node['wordpress']['languages']['themes'].to_a.each do |project|
    next unless urls[project]

    file = "#{node['wordpress']['dir']}/wp-content/themes/#{project}/languages/"
    file += "#{node['wordpress']['languages']['lang']}.mo"

    remote_file file do
      source urls[project]
      owner "root"
      group "root"
      mode "0644"
      action :create_if_missing
    end
  end
end


recipes/default.rb
#
# Cookbook Name:: wordpress
# Recipe:: default
#
# Copyright 2015, Cloudenablers
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe "apache2"
include_recipe "mysql::server"
include_recipe "mysql::ruby"
include_recipe "php"
include_recipe "php::module_mysql"
include_recipe "apache2::mod_php5"

puts node['wordpress']['db']['user']
if node.has_key?("ec2")
  server_fqdn = node['ec2']['public_hostname']
else
  server_fqdn = "wordpress"
end

node.set_unless['wordpress']['db']['password'] = secure_password
node.set_unless['wordpress']['keys']['auth'] = secure_password
node.set_unless['wordpress']['keys']['secure_auth'] = secure_password
node.set_unless['wordpress']['keys']['logged_in'] = secure_password
node.set_unless['wordpress']['keys']['nonce'] = secure_password


if node['wordpress']['version'] == 'latest'
  # WordPress.org does not provide a sha256 checksum, so we'll use the sha1 they do provide
  require 'digest/sha1'
  require 'open-uri'
  local_file = "#{Chef::Config[:file_cache_path]}/wordpress-latest.tar.gz"
  remote_file "#{Chef::Config[:file_cache_path]}/wordpress-latest.tar.gz" do
    source "http://wordpress.org/latest.tar.gz"
    mode "0644"
  end
else
  remote_file "#{Chef::Config[:file_cache_path]}/wordpress-#{node['wordpress']['version']}.tar.gz" do
    source "#{node['wordpress']['repourl']}/wordpress-#{node['wordpress']['version']}.tar.gz"
    mode "0644"
  end
end

directory node['wordpress']['dir'] do
  owner "root"
  group "root"
  mode "0755"
  action :create
  recursive true
end

execute "untar-wordpress" do
  cwd node['wordpress']['dir']
  command "tar --strip-components 1 -xzf #{Chef::Config[:file_cache_path]}/wordpress-#{node['wordpress']['version']}.tar.gz"
  creates "#{node['wordpress']['dir']}/wp-settings.php"
end

execute "mysql-install-wp-privileges" do
  command "/usr/bin/mysql -u root -h \"#{node['wordpress']['db']['host']}\" -p\"#{node['mysql']['server_root_password']}\" < #{node['mysql']['conf_dir']}/wp-grants.sql"
  action :nothing
end

template "#{node['mysql']['conf_dir']}/wp-grants.sql" do
  source "grants.sql.erb"
  owner "root"
  group "root"
  mode "0600"
  variables(
    :user     => node['wordpress']['db']['user'],
    :password => node['mysql']['server_root_password'],
    :database => node['wordpress']['db']['database'],
    :host	=> server_fqdn
  )
  notifies :run, "execute[mysql-install-wp-privileges]", :immediately
end

execute "create #{node['wordpress']['db']['database']} database" do
  command "/usr/bin/mysqladmin -u root -h \"#{node['wordpress']['db']['host']}\" -p\"#{node['mysql']['server_root_password']}\" create #{node['wordpress']['db']['database']}"
  not_if do
    # Make sure gem is detected if it was just installed earlier in this recipe
    require 'rubygems'
    Gem.clear_paths
    require 'mysql'
    m = Mysql.new(node['wordpress']['db']['host'], "root", node['mysql']['server_root_password'])
    m.list_dbs.include?(node['wordpress']['db']['database'])
  end
  notifies :create, "ruby_block[save node data]", :immediately unless Chef::Config[:solo]
end

# save node data after writing the MYSQL root password, so that a failed chef-client run that gets this far doesn't cause an unknown password to get applied to the box without being saved in the node data.
unless Chef::Config[:solo]
  ruby_block "save node data" do
    block do
      node.save
    end
    action :create
  end
end

log "wordpress_install_message" do
  action :nothing
  message "Navigate to 'http://#{server_fqdn}/wp-admin/install.php' to complete wordpress installation"
end

template "#{node['wordpress']['dir']}/wp-config.php" do
  source "wp-config.php.erb"
  owner "root"
  group "root"
  mode "0644"
  variables(
    :database        => node['wordpress']['db']['database'],
    :user            => node['wordpress']['db']['user'],
    :password        => node['mysql']['server_root_password'],
    :host			 => node['wordpress']['db']['host'],
    :auth_key        => node['wordpress']['keys']['auth'],
    :secure_auth_key => node['wordpress']['keys']['secure_auth'],
    :logged_in_key   => node['wordpress']['keys']['logged_in'],
    :nonce_key       => node['wordpress']['keys']['nonce'],
    :lang            => node['wordpress']['languages']['lang']
  )
  notifies :write, "log[wordpress_install_message]"
end

apache_site "000-default" do
  enable false
end

web_app "wordpress" do
  template "wordpress.conf.erb"
  docroot node['wordpress']['dir']
  server_name server_fqdn
  server_aliases node['wordpress']['server_aliases']
end

execute "disable default webiste" do
  command "/usr/sbin/a2dissite 000-default.conf"
end

webserver_name = 'apache2'
case node['platform']
when 'redhat', 'centos', 'scientific', 'fedora', 'suse', 'amazon', 'oracle'
	webserver_name = 'httpd'
end

execute "disable default webiste" do
  command "service #{webserver_name} reload"
end
